### DISCUSSION TITLE
How to think about `createWithEqualityFn` vs `create`

### DISCUSSION THREAD

**Discussion by charkour** [pmndrs/zustand#2243] [Q&A] [answered]
Hello, I have a question about how to think about `createWithEqualityFn` vs `create`.

In v4, `create` and `useStore` both had the function signature with `equalityFn` deprecated in favor of `createWithEqualityFn` and `useStoreWithEqualityFn`. Are the `WithEqualityFn` versions a temporary holdover that will be deprecated in a future version, or will they stick around and be an alternative way for zustand users to provide custom equality functions to their selectors? 

To avoid the deprecations in `create` and `useStore`, should I move to use the `WithEqualityFn`, or should I rewrite my app to have more granular store selectors that don't need an equality function?

Thanks

**Comment by dai-shi** [comment, marked as answer]
I originally thought that `WithEqualityFn` versions are a long-term solution. The point was separating them from normal ones, and not discouraging `equalityFn`.

However, we got some feedback after the release and our findings are:
- Most use cases of `equalityFn` would be `shallow` and we now provides `useShallow` that can be used with normal ones. (my rough estimation, 95% cases of `WithEqualityFn`)
- For the rest of use cases, it might be meaningful to rewrite for better performance. (4%)
- Otherwise, we could document how to use `use-sync-external-store` with vanilla  zustand store. (1%)

So, if that assumption are correct and the migration goes well, we may deprecate `WithEqualityFn` in the future.

**Reply by charkour** [reply]
Make sense. That is helpful context to know! Thank you
